Meta/lint-keymaps.py

```diff
@@ -0,0 +1,128 @@
+        if len(c) > 1:
+            report(filename, 'more than one character ("{}") for charmap index {} of {}'.format(c, i, mapname))
+            all_good = False
+
+    # TODO: Require that a few keys are set?
+
+    if len(values) not in GOOD_MAP_LENGTHS:
+        report(filename, 'length {} of map {} is suspicious. Off-by-one?'.format(len(values), mapname))
+        all_good = False
+
+    return all_good
+
+
+def validate_fullmap(filename, fullmap):
+    all_good = True
+
+    if not isinstance(fullmap, dict):
+        report(filename, 'is not an object')
+        return False  # Cannot continue other checks
+
+    for name, map_ in fullmap.items():
+        if name not in PERMITTED_MAPS:
+            report(filename, 'contains unknown entry {}'.format(name))
+            all_good = False
+
+        all_good &= validate_single_map(filename, name, map_)
+
+    for name in REQUIRED_MAPS:
+        if name not in fullmap:
+            report(filename, 'map {} is missing'.format(name))
+            all_good = False
+
+    if 'altgr_map' in fullmap and 'alt_map' in fullmap and fullmap['altgr_map'] == fullmap['alt_map']:
+        report(filename, 'altgr_map is identical to alt_map. Remove altgr_map for the same effect.')
+        report(filename, '(Or add new characters!)')
+        all_good = False
+
+    if 'shift_altgr_map' in fullmap and 'alt_map' in fullmap and fullmap['shift_altgr_map'] == fullmap['alt_map']:
+        report(filename, 'shift_altgr_map is identical to alt_map. Remove shift_altgr_map for the same effect.')
+        report(filename, '(Or add new characters!)')
+        all_good = False
+
+    return all_good
+
+
+def run_with(filenames):
+    passed = 0
+    for filename in filenames:
+        with open(filename, 'r') as fp:
+            fullmap = json.load(fp)
+        if validate_fullmap(filename, fullmap):
+            passed += 1
+
+    print('{} out of {} keymaps passed.'.format(passed, len(filenames)))
+    return passed == len(filenames)
+
+
+def list_files_here():
+    filelist = []
+    for filename in os.listdir():
+        if filename.endswith('.json'):
+            filelist.append(filename)
+        else:
+            report(filename, 'weird filename (ignored)')
+    # Files are in "filesystem" order. Sort them for slightly more
+    # aesthetically pleasing output.
+    filelist.sort()
+    return filelist
+
+
+def run_here():
+    return run_with(list_files_here())
+
+
+def display_for(filename, index):
+    with open(filename, 'r') as fp:
+        fullmap = json.load(fp)
+    for name in PERMITTED_MAPS:
+        c = None
+        if name in fullmap:
+            m = fullmap[name]
+            if len(m) > index and m[index]:
+                c = m[index]
+        if c is None:
+            print('{}: None'.format(name))
+        else:
+            print('{}: "{}"'.format(name, c))
+
+
+if __name__ == '__main__':
+    os.chdir(os.path.dirname(__file__) + "/../Base/res/keymaps/")
```

Comment:
No change required as far as I'm concerned, this is just a linting script... but you should try [`pathlib`](https://docs.python.org/3/library/pathlib.html) if you haven't yet :)

Reply:
It doesn't really help with going from `__file__` to the containing directory – at best it's another function. Simpler than addition is impossible (although I really like their `/` operator!), and `chdir` is a OS-level function, I don't really like to see it wrapped. So I'm sure `pathlib` is nice, but it solves a different problem :)

Reply:
You shouldn't chdir in the first place, listdir (and the pathlib equivalent) works with explicitly given paths as well. Not here to argue though, if it works, it works